File: mobile_client/lib/services/room_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:mobile_client/constants/endpoint_constants.dart';
import 'package:mobile_client/constants/enum/game_mode.dart';
import 'package:mobile_client/constants/enum/game_start_error_types.dart';
import 'package:mobile_client/constants/enum/page_url.dart';
import 'package:mobile_client/constants/enum/room_access_status.dart';
import 'package:mobile_client/constants/errors/game_starter_error.dart';
import 'package:mobile_client/constants/socket_events/room_event.dart';
import 'package:mobile_client/models/environment.dart';
import 'package:mobile_client/models/game_config_dto.dart';
import 'package:mobile_client/models/join_configs.dart';
import 'package:mobile_client/services/websocket_service.dart';
import 'package:mobile_client/utils/app_navigator.dart';

class RoomService extends ChangeNotifier {
  String roomId = '';
  bool isGameMaster = false;
  bool isLocked = false;
  GameMode gameMode = GameMode.classic;
  String waitingRoomBaseUrl = '${Environment.apiBaseUrl}/$waitingRoomEndpoint';
  ValueNotifier<bool> isProcessing = ValueNotifier(false);
  bool arePowerUpsEnabled = false;

  final WebSocketService _webSocketService;

  RoomService(this._webSocketService);

  resetManager() {
    roomId = '';
    isGameMaster = false;
    isLocked = false;
    gameMode = GameMode.classic;
    isProcessing.value = false;
    arePowerUpsEnabled = false;
  }

  Future<RoomAccessStatus?> canPlayerJoinRoom(String roomId) async {
    final response =
        await _webSocketService.sendWithAck(RoomEvent.canJoinRoom.name, roomId);
    return RoomAccessStatus.values.firstWhere((e) => e.name == response);
  }

  Future<void> createRoomAndJoin(
      GameConfigDto gameConfig, String? socketId) async {
    if (socketId == null) {
      throw Exception('Creator Socket ID cannot be null');
    }

    final response = await http.post(Uri.parse(waitingRoomBaseUrl),
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(
            {"gameConfig": gameConfig.toJson(), "creatorId": socketId}));

    if (response.statusCode == 201) {
      isGameMaster = true;
      gameMode = gameConfig.gameMode;
      roomId = jsonDecode(response.body)["roomId"];
      await _setupAndJoin();
    } else if (response.statusCode == 400 &&
        jsonDecode(response.body)['message'] ==
            "Organizer does not have enough money") {
      throw GameStartError(GameStartErrorTypes.notEnoughMoney);
    } else {
      throw Exception("Failed to create room: ${response.body}");
    }
  }

  Future<void> playerJoinRoom(String roomId) async {
    this.roomId = roomId;
    isGameMaster = false;

    await _setupAndJoin();
  }

  navigateToGamePage() {
    if (isGameMaster && gameMode != GameMode.elimination) {
      AppNavigator.go('/${PageUrl.gameMaster.name}');
    } else {
      AppNavigator.go('/${PageUrl.game.name}');
    }
  }

  Future<void> _setupAndJoin() async {
    isLocked = false;
    if (!_webSocketService.isSocketConnected()) return;

    final data = await _webSocketService.sendWithAck<String, dynamic>(
        RoomEvent.joinRoom.name, roomId);
    arePowerUpsEnabled = JoinConfigs.fromJson(data).arePowerUpsEnabled;
    gameMode = JoinConfigs.fromJson(data).gameMode;
    AppNavigator.go('/${PageUrl.waitingRoom.name}');
  }
}

File: desktop-client/src/app/home/services/room/room.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { SERVER_URL_API } from '@app/constants/server-url-and-api-constant';
import { PageUrl } from '@app/enum/page-url';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { WAITING_ROOM_ENDPOINT } from '@common/constants/endpoint-constants';
import { GameMode } from '@common/enum/game-mode';
import { RoomAccessStatus } from '@common/enum/room-access-status';
import { GameConfigDto } from '@common/interfaces/game-config-dto';
import { JoinConfigs } from '@common/interfaces/join-configs';
import { RoomId } from '@common/interfaces/room-id';
import { RoomEvent } from '@common/socket-events/room-event';
import { firstValueFrom } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class RoomService {
    roomId: string;
    isGameMaster: boolean = false;
    isLocked: boolean = false;
    gameMode: GameMode;
    baseUrl = `${SERVER_URL_API}/${WAITING_ROOM_ENDPOINT}`;
    arePowerUpsEnabled: boolean = false;

    // eslint-disable-next-line max-params
    constructor(
        private readonly http: HttpClient,
        private readonly router: Router,
        private readonly webSocketService: WebsocketService,
    ) {}

    resetManager() {
        this.roomId = '';
        this.isGameMaster = false;
        this.isLocked = false;
        this.gameMode = GameMode.CLASSIC;
        this.arePowerUpsEnabled = false;
    }

    async canPlayerJoinRoom(roomId: string): Promise<RoomAccessStatus> {
        return this.webSocketService.sendWithAck(RoomEvent.CAN_JOIN_ROOM, roomId);
    }

    async createRoomAndJoin(gameConfig: GameConfigDto, creatorSocketId: string): Promise<void> {
        const { roomId } = await firstValueFrom(this.http.post<RoomId>(this.baseUrl, { gameConfig, creatorId: creatorSocketId }));
        this.roomId = roomId;
        this.isGameMaster = true;
        this.gameMode = gameConfig.gameMode;

        await this.setupAndJoin();
    }

    async playerJoinRoom(roomId: string) {
        this.roomId = roomId;
        this.isGameMaster = false;

        await this.setupAndJoin();
    }

    navigateToGamePage() {
        if (this.isGameMaster && this.gameMode !== GameMode.ELIMINATION) {
            this.router.navigate([`/${PageUrl.APP_PREFIX}/${PageUrl.GAME_MASTER}`]);
        } else {
            this.router.navigate([`/${PageUrl.APP_PREFIX}/${PageUrl.GAME}`]);
        }
    }

    private async setupAndJoin() {
        this.isLocked = false;

        const { gameMode, arePowerUpsEnabled } = await this.webSocketService.sendWithAck<string, JoinConfigs>(RoomEvent.JOIN_ROOM, this.roomId);
        this.gameMode = gameMode;
        this.arePowerUpsEnabled = arePowerUpsEnabled;

        this.router.navigate([`/${PageUrl.APP_PREFIX}/${PageUrl.WAITING_ROOM}`]);
    }
}
